### DISCUSSION TITLE
Generics over keywords

### DISCUSSION THREAD

**Discussion by jtschuster** [dotnet/csharplang#6477] [Language Ideas]
Rust announced earlier this year that they're looking into making functions that can be generic over their keywords like `async` or `const` so that a single implementation of a function can be invoked as blocking or async. I think this could be an interesting idea to investigate.

Rust blog post: https://blog.rust-lang.org/inside-rust/2022/07/27/keyword-generics.html

**Comment by HaloFour** [comment]
I know that there have been some prototypes to explore `const` or `value` generics in the runtime.  As for `async` generics, I'm not sure how such a beast would work as `async` is strictly a language feature and the runtime is entirely unaware of it.  It sounds like it would require the compiler to do some kind of template expansion which would be very different to how generics work in .NET today.
